File: windows/src/Elevate.Core/Coordination/PropagationWatcher.cs
using Elevate.Core.Models;
using Elevate.Core.Providers;

namespace Elevate.Core.Coordination;

/// <summary>Where one role has got to between "PIM says active" and "it works". Port of the Swift <c>PropagationState</c>.</summary>
public enum PropagationState
{
    /// <summary>Active, probed, not usable yet. Keep waiting.</summary>
    Propagating,

    /// <summary>A probe saw the access. This is the state worth telling the user about.</summary>
    Ready,

    /// <summary>Still not usable when the deadline passed. Say so, and say what usually explains it.</summary>
    Unconfirmed,

    /// <summary>
    /// Nothing here can be observed from the client — a directory-scoped Entra role, group
    /// ownership, a sign-in method whose token cannot be read. Not a failure; the role is as active
    /// as it was, we simply cannot add anything to that.
    /// </summary>
    Unobservable,
}

/// <summary>One role's propagation verdict. <see cref="Detail"/> explains the two states that are not plain progress.</summary>
public sealed record PropagationOutcome(RoleKey RoleKey, PropagationState State, string? Detail = null)
{
    /// <summary>Whether the caller can stop waiting on this role, whatever the answer was.</summary>
    public bool IsSettled => State != PropagationState.Propagating;
}

/// <summary>
/// Turns "the service says active" into "the access works", by asking each role's provider until it
/// answers yes or the deadline passes. Roles are independent: one that cannot be observed does not
/// hold up the rest, and one that is ready is reported the moment it is, so a caller can notify.
/// Port of the Swift <c>PropagationWatcher</c>.
/// </summary>
public sealed class PropagationWatcher(
    ActivationCoordinator coordinator,
    Func<TimeSpan, CancellationToken, Task>? delay = null)
{
    private readonly Func<TimeSpan, CancellationToken, Task> _delay = delay ?? Task.Delay;

    /// <summary>How long to wait before the first probe: a round trip costs more than it buys at t=0.</summary>
    public TimeSpan FirstInterval { get; init; } = TimeSpan.FromSeconds(5);

    /// <summary>The gap between probes settles here. Propagation is measured in minutes; polling faster only costs calls.</summary>
    public TimeSpan MaxInterval { get; init; } = TimeSpan.FromSeconds(20);

    /// <summary>
    /// Probes until every role has settled. <paramref name="deadline"/> defaults to the longest
    /// <see cref="PropagationHints.Deadline(RoleScopeKind)"/> among the roles. <paramref name="onChange"/>
    /// fires once per role, when it settles.
    /// </summary>
    public async Task<IReadOnlyList<PropagationOutcome>> WaitAsync(
        IReadOnlyList<ActiveAssignment> assignments,
        IEnumerable<Identity> identities,
        TimeSpan? deadline = null,
        Action<PropagationOutcome>? onChange = null,
        CancellationToken ct = default)
    {
        ArgumentNullException.ThrowIfNull(assignments);
        ArgumentNullException.ThrowIfNull(identities);

        var identityById = identities.DistinctBy(i => i.Id).ToDictionary(i => i.Id);
        var pending = assignments
            .Where(a => a.Status.Kind == AssignmentStatusKind.Active)
            .DistinctBy(a => a.RoleKey)
            .ToList();

        var settled = new Dictionary<RoleKey, PropagationOutcome>();
        foreach (var assignment in pending.Where(a => !identityById.ContainsKey(a.RoleKey.IdentityId)))
        {
            settled[assignment.RoleKey] = Settle(onChange, new PropagationOutcome(
                assignment.RoleKey, PropagationState.Unobservable, "That account is not signed in here."));
        }

        pending.RemoveAll(a => settled.ContainsKey(a.RoleKey));
        if (pending.Count == 0)
        {
            return [.. settled.Values];
        }

        var until = DateTimeOffset.UtcNow
            + (deadline ?? PropagationHints.Deadline(pending.Select(a => a.RoleKey.Scope.Kind)));
        var interval = FirstInterval;
        while (pending.Count > 0)
        {
            await _delay(interval, ct).ConfigureAwait(false);
            interval = interval < MaxInterval ? MaxInterval : interval;

            var round = await Task.WhenAll(pending.Select(a => ProbeAsync(a, identityById[a.RoleKey.IdentityId], ct)))
                .ConfigureAwait(false);

            var expired = DateTimeOffset.UtcNow >= until;
            foreach (var (assignment, access) in pending.Zip(round))
            {
                var outcome = Verdict(assignment.RoleKey, access, expired);
                if (outcome is not null)
                {
                    settled[assignment.RoleKey] = Settle(onChange, outcome);
                }
            }

            pending.RemoveAll(a => settled.ContainsKey(a.RoleKey));
        }

        return [.. settled.Values];
    }

    /// <summary>The outcome a probe settles on, or null while the role is still worth asking about.</summary>
    private static PropagationOutcome? Verdict(RoleKey key, EffectiveAccess access, bool expired) => access.Kind switch
    {
        EffectiveAccessKind.Confirmed => new PropagationOutcome(key, PropagationState.Ready),
        EffectiveAccessKind.Unknown => new PropagationOutcome(key, PropagationState.Unobservable, access.Detail),
        _ when expired => new PropagationOutcome(
            key, PropagationState.Unconfirmed, PropagationHints.LikelyCause(key.Scope.Kind)),
        _ => null,
    };

    private async Task<EffectiveAccess> ProbeAsync(ActiveAssignment assignment, Identity identity, CancellationToken ct)
    {
        if (coordinator.Provider(assignment.RoleKey.Scope.Kind) is not { } provider)
        {
            return EffectiveAccess.Unknown("No provider for this kind of role.");
        }

        try
        {
            return await provider.EffectiveAccessAsync(assignment, identity, ct).ConfigureAwait(false);
        }
        catch (PimException e)
        {
            // A refused or failed probe is not a failed activation: the role is active either way.
            return EffectiveAccess.Unknown(e.Message);
        }
        catch (Exception e) when (e is not OperationCanceledException)
        {
            return EffectiveAccess.Unknown(e.Message);
        }
    }

    private static PropagationOutcome Settle(Action<PropagationOutcome>? onChange, PropagationOutcome outcome)
    {
        onChange?.Invoke(outcome);
        return outcome;
    }
}

File: macos/Sources/ElevateCore/Coordination/PropagationWatcher.swift
import Foundation

/// Where one role has got to between "PIM says active" and "it works".
public enum PropagationState: Hashable, Sendable {
    /// Active, probed, not usable yet. Keep waiting.
    case propagating
    /// A probe saw the access. This is the state worth telling the user about.
    case ready
    /// Still not usable when the deadline passed. Say so, and say what usually explains it.
    case unconfirmed
    /// Nothing here can be observed from the client — a directory-scoped Entra role, group
    /// ownership, a sign-in method whose token cannot be read. Not a failure; the role is as active
    /// as it was, we simply cannot add anything to that.
    case unobservable
}

/// One role's propagation verdict. `detail` explains the two states that are not plain progress.
public struct PropagationOutcome: Hashable, Sendable {
    public let roleKey: RoleKey
    public let state: PropagationState
    public let detail: String?

    public init(roleKey: RoleKey, state: PropagationState, detail: String? = nil) {
        self.roleKey = roleKey
        self.state = state
        self.detail = detail
    }

    /// Whether the caller can stop waiting on this role, whatever the answer was.
    public var isSettled: Bool { state != .propagating }
}

/// Turns "the service says active" into "the access works", by asking each role's provider until it
/// answers yes or the deadline passes. Roles are independent: one that cannot be observed does not
/// hold up the rest, and one that is ready is reported the moment it is, so a caller can notify.
public final class PropagationWatcher: Sendable {
    /// How long to wait before the first probe: a round trip costs more than it buys at t=0.
    public static let defaultFirstInterval: TimeInterval = 5
    /// The gap between probes settles here. Propagation is measured in minutes; polling faster only costs calls.
    public static let defaultMaxInterval: TimeInterval = 20

    public let firstInterval: TimeInterval
    public let maxInterval: TimeInterval
    private let coordinator: ActivationCoordinator
    private let sleep: @Sendable (TimeInterval) async throws -> Void

    public init(
        coordinator: ActivationCoordinator,
        firstInterval: TimeInterval = PropagationWatcher.defaultFirstInterval,
        maxInterval: TimeInterval = PropagationWatcher.defaultMaxInterval,
        sleep: @escaping @Sendable (TimeInterval) async throws -> Void = { try await Task.sleep(for: .seconds($0)) }
    ) {
        self.coordinator = coordinator
        self.firstInterval = firstInterval
        self.maxInterval = maxInterval
        self.sleep = sleep
    }

    /// Probes until every role has settled. `deadline` defaults to the longest
    /// `PropagationHints.deadline` among the roles. `onChange` fires once per role, when it settles.
    @discardableResult
    public func wait(
        for assignments: [ActiveAssignment],
        identities: [Identity],
        deadline: TimeInterval? = nil,
        onChange: (@Sendable (PropagationOutcome) -> Void)? = nil
    ) async throws -> [PropagationOutcome] {
        let identityById = Dictionary(identities.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })
        var pending = assignments.filter { $0.status == .active }
        var seen = Set<RoleKey>()
        pending = pending.filter { seen.insert($0.roleKey).inserted }

        var settled: [RoleKey: PropagationOutcome] = [:]
        for assignment in pending where identityById[assignment.roleKey.identityId] == nil {
            let outcome = PropagationOutcome(
                roleKey: assignment.roleKey, state: .unobservable, detail: "That account is not signed in here.")
            onChange?(outcome)
            settled[assignment.roleKey] = outcome
        }
        pending.removeAll { settled[$0.roleKey] != nil }
        guard !pending.isEmpty else { return Array(settled.values) }

        let until = Date.now.addingTimeInterval(deadline ?? PropagationHints.deadline(pending.map(\.roleKey.scope.kind)))
        var interval = firstInterval
        while !pending.isEmpty {
            try await sleep(interval)
            interval = maxInterval

            var answers: [(ActiveAssignment, EffectiveAccess)] = []
            for assignment in pending {
                guard let identity = identityById[assignment.roleKey.identityId] else { continue }
                answers.append((assignment, await probe(assignment, identity: identity)))
            }

            let expired = Date.now >= until
            for (assignment, access) in answers {
                if let outcome = Self.verdict(assignment.roleKey, access: access, expired: expired) {
                    onChange?(outcome)
                    settled[assignment.roleKey] = outcome
                }
            }
            pending.removeAll { settled[$0.roleKey] != nil }
        }
        return Array(settled.values)
    }

    /// The outcome a probe settles on, or nil while the role is still worth asking about.
    static func verdict(_ key: RoleKey, access: EffectiveAccess, expired: Bool) -> PropagationOutcome? {
        switch access {
        case .confirmed:
            PropagationOutcome(roleKey: key, state: .ready)
        case let .unknown(reason):
            PropagationOutcome(roleKey: key, state: .unobservable, detail: reason)
        case .notYet:
            expired
                ? PropagationOutcome(roleKey: key, state: .unconfirmed, detail: PropagationHints.likelyCause(key.scope.kind))
                : nil
        }
    }

    private func probe(_ assignment: ActiveAssignment, identity: Identity) async -> EffectiveAccess {
        guard let provider = coordinator.provider(for: assignment.roleKey.scope.kind) else {
            return .unknown("No provider for this kind of role.")
        }
        do {
            // A refused or failed probe is not a failed activation: the role is active either way.
            return try await provider.effectiveAccess(assignment, identity: identity)
        } catch let error as PIMError {
            return .unknown(error.userMessage)
        } catch {
            return .unknown(error.localizedDescription)
        }
    }
}
